### PR TITLE
Server 2297/externalizing pg mongo

### PR DESCRIPTION
**Pull request by soulchips**
# Description
Adding instructions on how to externalize your postgres and mongodb instances in server 3.x - 4.x

# Reasons
https://circleci.atlassian.net/browse/SERVER-2297

# Content Checklist
Please follow our style when contributing to CircleCI docs. Our style guide is here: [https://circleci.com/docs/style/style-guide-overview](https://circleci.com/docs/style/style-guide-overview).

Please take a moment to check through the following items when submitting your PR (this is just a guide so will not be relevant for all PRs) 😸:

- [ ] Break up walls of text by adding paragraph breaks.
- [ ] Consider if the content could benefit from more structure, such as lists or tables, to make it easier to consume.
- [ ] Keep the title between 20 and 70 characters.
- [ ] Consider whether the content would benefit from more subsections (h2-h6 headings) to make it easier to consume.
- [ ] Check all headings h1-h6 are in sentence case (only first letter is capitalized).
- [ ] Is there a "Next steps" section at the end of the page giving the reader a clear path to what to read next?
- [ ] Include relevant backlinks to other CircleCI docs/pages.

### CODE REVIEW

**Comment by atulsingh0** [comment]
LGTM, One query though, Don't we want to share the supported mongo-db version as well and to do a step-up migration (as done by @ryanwohara) from existing CCI instance.

**Review by atulsingh0 — APPROVED** [review]

**Comment by rosieyohannan** [comment]
@soulchips thanks for this! Just checking - I've read through and I think the steps for each server version are exactly the same? If so I will create a reusable snippet to be included multiple times if that's ok? Just to make keeping this up to date more straightforward. If there are subtle differences though this won't work as well. Let me know what you think.

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 1]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
 * Using encrypted storage for backups as databases might contain sensitive information.
 * Performing a backup before each upgrade of CircleCI server.
 
+[#migrating-from-internal-postgres]
+=== Migrating from an Internal Postgres to an Externalized Source
+
+NOTE: If you are doing a fresh install of CircleCI server, then you can skip this section and head to <<connecting-your-external-postgres>>
+
+When a CircleCI Server instance deployed, Postgres is deployed internally by default via its helm chart. However as an operator, you may wish to externalize this database to have better control over it's scalability and availability. Once you have configured your external Postgres, you may use the guide below to migrate your Postgres data to your external database.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:88]:
```suggestion
When a CircleCI server instance is deployed, by default Postgres is deployed within the cluster via its helm chart. However as an operator, you may wish to externalize this database to have better control over it's scalability and availability. Once you have configured your external Postgres, you may use the guide below to migrate your Postgres data to your external database.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 2]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+
+NOTE: This process requires downtime.
+
+==== 1. Disable the application
+We disable the CircleCI Server application by scaling down the application layer pods using the following command. No Data is lost in this process, but the application will be unreachable.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:93]:
```suggestion
We disable the CircleCI server application by scaling down the application layer pods using the following command. No Data is lost in this process, but the application will be unreachable.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 3]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+[source,shell]
+----
+namespace=<your-server-namespace>
+kubectl -n "$namespace" scale deploy -l "layer=application" --replicas="0"
+----
+
+Running `kubectl -n "$namespace" get pods` should show most of your pods scaling to down, leaving your database pods running including postgres.
+
+==== 2. Validate access to your External Postgres from within the Cluster (Optional)
+Here we will confrim that pods within your CircleCI Server cluster can access your external Postgres. We will do this from within your internal Postgres.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:105]:
```suggestion
Here we will confrim that pods within your CircleCI server cluster can access your external Postgres. We will do this from within your internal Postgres.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 4]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+Retrieve your internal Postgres credentials:
+
+[source,shell]
+----
+PG_PASSWORD=$(kubectl -n "$namespace" get secrets postgresql -o jsonpath="{.data.postgresql-password}" | base64 --decode)
+----
+
+NOTE: The username for your internal Postgres is `postgres`. The password is randomly generated unless directly set at installation.
+
+Connect to Postgres pod and perform postgres dump
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:132]: There are some inconsistencies with `Postgres` and `postgres` in a few spots

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 5]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+
+==== 4. Restore your data in your External Postgres
+While still connected to your the internally deployed Postgres, we will restore the dumped data to your external postgres
+
+[source,shell]
+----
+psql -h <your-external-postgres-host> -U postgres -p <your-external-postgres-port> < circle.sql
+----
+
+Now your external Postgres will have your CircleCI Server data. In the next section we will update CircleCI server to Point to your external Postgres.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:163]:
```suggestion
Now your external Postgres will have your CircleCI server data. In the next section we will update CircleCI server to Point to your external Postgres.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 6]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+
+* *PostgreSQL Service Domain (required)* - The domain or IP address of your PostgreSQL instance. 
+
+* *PostgreSQL Service Port (required)* - The port of your PostgreSQL instance. 
+
+* *PostgreSQL Service Username (required)* - A user with the appropriate privileges to access your PostgreSQL instance. 
+
+* *PostgreSQL Service Password (required)* - The password of the user account used to access your PostgreSQL instance.
+
+Once you apply your changes in KOTS admin, your CircleCI Server instance will be restored with pods running as before.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:179]:
```suggestion
Once you apply your changes in KOTS admin, your CircleCI server instance will be restored with pods running as before.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 7]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+
+[source,shell]
+----
+psql -h <your-external-postgres-host> -U postgres -p <your-external-postgres-port> < circle.sql
+----
+
+Now your external Postgres will have your CircleCI Server data. In the next section we will update CircleCI server to Point to your external Postgres.
+
+[#connecting-your-external-postgres]
+=== Connecting Your External Postgres Instance to CircleCI Server [[connecting-your-external-postgres]]
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:166]:
```suggestion
=== Connecting your external Postgres instance to CircleCI server [[connecting-your-external-postgres]]
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 8]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+
 == MongoDB
 
 NOTE: If using your own MongoDB instance, it needs to be version 3.6 or higher. 
 
-.External MongoDB
+[#migrating-from-internal-mongodb]
+=== Migrating from an Internal MongoDB to an Externalized Source
+
+NOTE: If you are doing a fresh install of CircleCI Server, then you can skip this section and head to <<connecting-your-external-mongodb>>
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:189]:
```suggestion
NOTE: If you are doing a fresh install of CircleCI server, then you can skip this section and head to <<connecting-your-external-mongodb>>
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 9]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+
+When a CircleCI Server instance deployed, MongoDB is deployed internally by default via its helm chart. However as an operator, you may wish to externalize this database to have better control over it's scalability and availability. Once you have configured your external MongoDB, you may use the guide below to migrate your Mongo data to your external database.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:191]:
```suggestion
When a CircleCI server instance deployed, MongoDB is deployed internally by default via its helm chart. However as an operator, you may wish to externalize this database to have better control over it's scalability and availability. Once you have configured your external MongoDB, you may use the guide below to migrate your MongoDB data to your external database.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 10]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+
+NOTE: This process requires downtime.
+
+==== 1. Disable the application
+We disable the CircleCI Server application by scaling down the application layer pods using the following command. No Data is lost in this process, but the application will be unreachable.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:196]:
```suggestion
We disable the CircleCI server application by scaling down the application layer pods using the following command. No Data is lost in this process, but the application will be unreachable.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 11]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+[source,shell]
+----
+namespace=<your-server-namespace>
+kubectl -n "$namespace" scale deploy -l "layer=application" --replicas="0"
+----
+
+Running `kubectl -n "$namespace" get pods` should show most of your pods scaling to down, leaving your database pods running including mongo.
+
+==== 2. Validate access to your External MongoDB from within the Cluster (Optional)
+Here we will confrim that pods within your CircleCI Server cluster can access your external MongoDB. We will do this from within your internal MongoDB's pod.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:208]:
```suggestion
Here we will confrim that pods within your CircleCI server cluster can access your external MongoDB. We will do this from within your internal MongoDB's pod.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 12]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+We disable the CircleCI Server application by scaling down the application layer pods using the following command. No Data is lost in this process, but the application will be unreachable.
+Scale down your application layer pods:
+
+[source,shell]
+----
+namespace=<your-server-namespace>
+kubectl -n "$namespace" scale deploy -l "layer=application" --replicas="0"
+----
+
+Running `kubectl -n "$namespace" get pods` should show most of your pods scaling to down, leaving your database pods running including mongo.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:205]:
```suggestion
Running `kubectl -n "$namespace" get pods` should show most of your pods scaling to down, leaving your database pods running including MongoDB.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 13]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+We can now use the generated MongoDB backup to restore the data to your external MongoDB:
+[source,shell]
+----
+kubectl -n "$namespace" exec "$MONGO_POD" -- mongorestore --drop -u "$MONGODB_USERNAME" -p "$MONGODB_PASSWORD" --authenticationDatabase admin /tmp/backups/circle_ghe;
+----
+
+Now your external MongoDB will have your CircleCI Server data. In the next section we will update CircleCI server to Point to your external MongoDB.
+
+[#connecting-your-external-mongodb]
+=== Connecting Your External MongoDB Instance to CircleCI Server [[connecting-your-external-mongodb]]
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:257]:
```suggestion
=== Connecting your external MongoDB instance to CircleCI server [[connecting-your-external-mongodb]]
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 14]
```diff
@@ -126,6 +283,8 @@ The authentication mechanism to use, usually SCRAM-SHA-1.
 * *Additional connection options (optional)* - 
 Any other connection options you would like to use. This needs to be formatted as a query string (`key=value` pairs, separated by `&`. Special characters must be URL encoded). See the link:https://docs.mongodb.com/v3.6/reference/connection-string/[MongoDB docs] for available options.
 
+Once you apply your changes in KOTS admin, your CircleCI Server instance will be restored with pods running as before.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:286]:
```suggestion
Once you apply your changes in KOTS admin, your CircleCI server instance will be restored with pods running as before.
```

File: jekyll/_cci2/server-3-operator-externalizing-services.adoc [hunk 15]
```diff
@@ -93,11 +80,181 @@ We strongly recommend the following:
...
+----
+
+==== 4. Restore your data in your External MongoDB
+We can now use the generated MongoDB backup to restore the data to your external MongoDB:
+[source,shell]
+----
+kubectl -n "$namespace" exec "$MONGO_POD" -- mongorestore --drop -u "$MONGODB_USERNAME" -p "$MONGODB_PASSWORD" --authenticationDatabase admin /tmp/backups/circle_ghe;
+----
+
+Now your external MongoDB will have your CircleCI Server data. In the next section we will update CircleCI server to Point to your external MongoDB.
```
Comment (ryanwohara) [inline jekyll/_cci2/server-3-operator-externalizing-services.adoc:254]:
```suggestion
Now your external MongoDB will have your CircleCI Server data. In the next section we will update CircleCI server to point to your external MongoDB.
```

**Comment by rosieyohannan** [comment]
> > @soulchips thanks for this! Just checking - I've read through and I think the steps for each server version are exactly the same? If so I will create a reusable snippet to be included multiple times if that's ok? Just to make keeping this up to date more straightforward. If there are subtle differences though this won't work as well. Let me know what you think.
> 
> Ah yes. If we can do that it would be ideal!

**Comment by rosieyohannan** [comment]
closing because same changes published here: https://github.com/circleci/circleci-docs/pull/8040

**Pull request closed without merging**